### DISCUSSION TITLE
Should `loadConfig` use a plugin?

### DISCUSSION THREAD

**Discussion by patrick91** [pyscript/pyscript#1437] [Proposals]
See: https://github.com/pyscript/pyscript/blob/0a4e36ae098f4f1f1860f602f01c84e569cbb382/pyscriptjs/src/main.ts#LL44C1-L44C1

I'm thinking of trying to use pyscript without using the custom elements directly (from React) and I was wondering if the `loadConfig` function should use a plugin instead of trying to find the configuration from an element?

I haven't tried to use pyscript just yet, when I do I'll write an example of I think I want it to work with react 😊

**Comment by hoodmane** [comment]
I think this request is out of scope for Pyscript. If you don't like the behavior of the pyscript custom elements I'd recommend using Pyodide and writing your own custom elements. (There are some people already doing this.)

**Reply by patrick91** [reply]
I'm currently using Pyodide and it works well, but considering that pyscript will allow multiple runtimes (pyodide, cpython-wasm, micropython) I think it would be nice to have pyscript work in "headless" mode :)

**Reply by hoodmane** [reply]
I'm trying to make micropython into as close as I can get to a drop-in replacement for Pyodide (observe how small the changes to the typescript code are in #1392). If they have the same API then the logic for swapability in your own application will also be very simple.

**Comment by JeffersGlass** [comment]
When you say "loadConfig should use a plugin", do you mean you'd like it to be configurable/disable-able? Or that you'd pass some kind of existing object to loadConfig and use that as the configuration? I guess I'm a little lost as to what your goal is.

**Reply by patrick91** [reply]
> Or that you'd pass some kind of existing object to loadConfig and use that as the configuration?

yes, this :)

sorry for being unclear! 😊

I'd love to be able to use pyscript somewhat like this in future:

```tsx
// imports here

const ps = PyScript({ runtime: "micropython", worker: true });

await ps.load();

await ps.runPython("some code here");
```

**Reply by hoodmane** [reply]
I guess you could imagine three layers:
1. Pyodide / Micropython / whatever runtime
2. A project which does worker setup and abstracts over which Python runtime we use
3. A project which adds custom html tags

This would arguably be a good design.

**Reply by patrick91** [reply]
yup, that sounds good :)

This is what I have for pyodide: https://github.com/patrick91/absurd-django/blob/main/components/pyodide.tsx

I guess I could make generic if the apis between pyodide and micropython are similar, but I'd prefer not to duplicate any code done by you folks 😊 The py-config has already a nice way to define dependencies for example 😊

**Reply by WebReflection** [reply]
love the suggestion ... if I might expose my taste here:

```js
const ps = await PyScript({ runtime: "micropython", worker: true });

await ps.run("some code here");
```

I don't see the need for `await ps.load()` as I think in no circumstances a *PyScript* project is imported and never loaded/used (and we can anyway bootstrap it to make it as ready as possible whenever `run` is executed or use `run` to bootstrap first time it's invoked) + the only thing we can currently do is to evaluate *Python* so that `ps.run(...)` looks like already explicit enough on the intent 😇
